File: backend/semgrep_rules/tests/javascript/kisa_sw_07_untrusted_url_redirect.js
function unsafe(req, res) {
  const next = req.query.next;
  // ruleid: kisa.sw07.javascript.external-input-to-redirect
  res.redirect(next);
}
function safe(req, res) {
  // ok: kisa.sw07.javascript.external-input-to-redirect
  res.redirect("/home");
}

File: backend/semgrep_rules/tests/java/kisa_sw_07_untrusted_url_redirect.java
import javax.servlet.http.*;

class kisa_sw_07_untrusted_url_redirect {
    void unsafe(HttpServletRequest req, HttpServletResponse resp) throws Exception {
        String next = req.getParameter("next");
        // ruleid: kisa.sw07.java.external-input-to-redirect
        resp.sendRedirect(next);
    }
    void safe(HttpServletResponse resp) throws Exception {
        // ok: kisa.sw07.java.external-input-to-redirect
        resp.sendRedirect("/home");
    }
}
